### ISSUE TITLE
Filter flag combination for dependencies or dependents only?

### ISSUE THREAD

**Issue by morewry** [lerna/lerna#2345]
Not a bug. More of a potential opportunity in the semantics of filter flags or a feature request: a filter flag combination that enables operating on dependents or dependencies _**only**_.

## Expected Behavior

Given `lerna --include-dependencies --scope packageC run build` where `packageC` depends on `packageA` and `packageB`, all 3 will `run build`.

I have a scenario where I want to build the dependencies of `packageC` one time, then build and watch for changes on `packageC`. Unfortunately, `packageC` gets built twice, once when I run the command above, and again when I run `lerna --scope packageC run dev`&mdash;simply because by convention all the packages have a `build` script (build number 1) and the build tool runs an initial build when watching (build number 2). Some of the builds are involved (and therefore take long) enough that it improves developer experience to optimize away a second build.

That can probably be solved in other reasonable ways, but this isn't the first time I've wished for the ability to filter this way. Commonly it's when I've been working on `packageC` and now want to verify its dependents, `packageD` and `packageE`. While working on `packageC`, I've built it and verified it, so it's actually extraneous in `lerna --include-dependents --scope packageC run build && lerna --include-dependents --scope packageC run test`. 

## Current Behavior

I tried `npx lerna --include-dependents --scope packageC --ignore packageC` just in case it might work, but it errors as not matching any packages.

## Possible Solution

I think the semantics of a command with `--scope` and `--ignore` being set to the same package and an `--include-dependencies`/`--include-dependents` flag could reasonably be interpreted to match the dependencies only. But that may not be widely intuitive and I understand implementation details might preclude it; another flag like `--only-dependencies`/`--only-dependents` could also meet the need.

**Comment by felixhayashi** [comment]
Bump! I have the exact same requirement.

In a situation where one works with `create-react-app (cra)`, it is required to first ensure all dependencies are build before actually starting or building the react app.

> You need to transpile the packages and specify them correctly as dependencies. Webpack handles the rest without any special handling, as they are just npm modules.
> (@evocateur, https://github.com/lerna/lerna/issues/1156#issuecomment-441407468)

Now in my situation, during development, the main package (which would map to `packageC` mentioned in the OP) is executed via webpack and uses hot-reloading, however before the server is started I need to build all dependencies first while **excluding the current package** to avoid triggering (and waiting) for an unnecessary build of the main package (which handled by webpack).

Intuitively I used the same semantics as suggested in the OP (`npx lerna --include-dependents --scope main-app --ignore main-app`), however this did not work.

Of course, I could add all `build` commands manually and chain them via `&&` in my `app-main` package but this is redundant and error-prone :-1: 

I would rather prefer being able to do this :+1: 

```bash
npm run lerna -- build --include-dependents --scope main-app --ignore main-app
lerna run start --scope main-app
```

**Comment by morewry** [comment]
I actually have a similar case I didn't mention with a Storybook app in the same repo.

There is a non-zero chance I could contribute this given a few hints from someone already familiar with codebase regarding the best approach.

**Comment by evocateur** [comment]
Sounds like you need to tune your build, tbh. If the inputs (source code + dependency graph) haven't changed, why does the code need to be rebuilt? [`pectin`](https://www.npmjs.com/package/pectin) and [`tsc --build`](https://www.typescriptlang.org/docs/handbook/project-references.html) are two examples of this "laziness" in action.

**Comment by evocateur** [comment]
As for omitting the scope-targeted package, pipe `lerna ls` into `grep -v '^${SCOPE}$'` and there you go. For example, in the root of pectin's [monorepo](https://github.com/evocateur/pectin):
```sh
$ npx lerna ls --scope @pectin/core --include-dependents
@pectin/core
@pectin/api
rollup-config-pectin
pectin
# a lot of extra logging, but to stderr (stdout only contains the package names)

$ npx lerna ls --scope @pectin/core --include-dependents | grep -v '^@pectin/core$'
@pectin/api
rollup-config-pectin
pectin
```

You could output that list itself as a series of `--scope` flags for any filterable lerna command:
```sh
npx lerna run build $(npx lerna ls --scope main-app --include-dependencies | grep -v '^main-app$' | xargs -n1 echo '--scope')

# in isolation:
npx lerna ls --scope main-app --include-dependencies | grep -v '^main-app$' | xargs -n1 echo '--scope'
```
(when translating to npm scripts, remove the `npx ` prefix)

**Comment by felixhayashi** [comment]
@evocateur your suggestion to scope the `lerna ls` and then filter and apply the results via `grep` and `xargs` is a very neat idea, thanks :+1:  I will try to go with this suggestion. Since all our devs are working on linux/mac machines, this solution appears suitable.

> Sounds like you need to tune your build, tbh. If the inputs (source code + dependency graph) haven't changed, why does the code need to be rebuilt?

in our case, there is no mechanism in the monorepo to detect what packages' code has changed. A developer may touch the code of several lib-packages to complete a task and therefore compiling all depdencies before starting the webpack server of the "main" project is a simple way to make sure every dependency is build and uptodate.

**Comment by morewry** [comment]
Yeah, the tools I am using already don't work that way unless you've already started a change watcher. I think Lerna should be tool agnostic here, but the `xargs` thing probably will work, though my npm script entry points are already more bash-complex than they should be.

**Comment by morewry** [comment]
Yeah, but the tools I am using already don't work that way unless you've already started a change watcher. They do once you have. In all ways other than these two build cases that's sufficient. I think Lerna should be tool agnostic here.

Pectin might help from a skim of its readme, so thanks for that tip. But builds and change detection of rebuilds aren't my only use case, as I mentioned. 

For all three cases I have, the `xargs` thing probably will work, though it's discouraging because my npm script entry points are already more bash-complex than they should be.

**Comment by shrugs** [comment]
in my experience the `--ignore` option isn't relevant when using `--include-dependencies`, though i'd like it to be

in my case, a storybook project with a component library exists in the monorepo. to build my project in CI I run something like `lerna run --scope @org/app --include-dependencies build`, which will also run the build-storybook since the component library `@org/components` is a dependency of `@org/app`. would be cool if the command `lerna run --scope @org/app --include-dependencies --ignore @org/components build` correctly ignored the build script for `@org/components`

here are the logs from my lerna command - the scopes look nice, but will continue to build `@org/components`

```
$ lerna run --include-dependencies --stream build --scope=@org/app --ignore @org/components
lerna notice cli v3.22.1
lerna notice filter including "@org/app"
lerna notice filter excluding "@org/components"
lerna notice filter including dependencies
lerna info filter [ '@org/app', '!@org/components' ]
lerna info Executing command in 3 packages: "yarn run build"
```

**Comment by schmod** [comment]
this is the shell script I came up with:

```sh
PKG=$1
LERNA=$(yarn bin lerna)

$LERNA ls -a -p --scope $PKG --include-dependencies --json \
  | jq -r '.[].name | select(contains("'$PKG'") | not) | "--scope " + .' \
  | xargs -t $LERNA run build
```

**Comment by tetslee** [comment]
I needed a cross platform solution:

// lerna-dependencies-only.js
```javascript
const { execSync } = require('child_process');
const { detectProjects } = require('lerna/utils');
const cliArgs = process.argv.slice(2);
const run = async () => {
  if (!cliArgs.length) {
    console.warn('no packages specified');
    return;
  }
  const { projectGraph } = await detectProjects();
  const isPackage = arg => projectGraph.dependencies[arg];
  const packages = cliArgs.filter(isPackage);
  const lernaArgs = cliArgs.filter(x => !isPackage(x));
  const depsSet = packages
    .flatMap(p => projectGraph.dependencies[p])
    .reduce((acc, dep) => {
      acc[dep.target] = true;
      return acc;
    }, {});
  const command = `npx lerna ${lernaArgs.join(' ')} ${Object.keys(depsSet)
    .map(d => `--scope="${d}"`)
    .join(' ')}`;
  console.log(command);
  execSync(command, { stdio: 'inherit' });
};
run();
```

Usage:
`node lerna-dependencies-only.js [lerna command args] [...packages]`
e.g.
`node lerna-dependencies-only.js run build --stream package1 package2`
